### ISSUE TITLE
Moderate-size grain backups getting truncated

### ISSUE THREAD

**Issue by mrdomino** [sandstorm-io/sandstorm#2726]
This is happening on one of my instances, and not another. The instance with the problem is an Ubuntu Xenial VM hosted on scaleway.com. The good one is a Debian VPS run by a friend.

I'm seeing different file sizes. Typically it gets to around 500KB the first time. It gets farther if it's retried multiple times in a row -- 3 to 4MB -- but still not all the way except rarely.

**Comment by zarvox** [comment]
Anything getting logged in syslog/dmesg?  I'd check that you're not hitting out-of-memory conditions, where the zip process generating the backup would be the process selected for the OOM killer to terminate.

**Comment by mrdomino** [comment]
It's not memory -- the VPS has 2GB and only 200MB is in use. I'll check dmesg next chance I get.

**Comment by mrdomino** [comment]
Nothing suspicious in dmesg or syslog either.

**Comment by zarvox** [comment]
Thanks for looking.  Is there anything in the admin log (Admin panel -> System log, or `/admin/status`) or grain log (computer monitor button) that might tell us anything else about where or how this is failing?

**Comment by mrdomino** [comment]
Nope, sorry. Everything looks boring and ordinary there.

**Comment by JamborJan** [comment]
Hey guys, I'm seeing the same issue on multiple sandstorm installations. It's happening for apps of any kind as soon as the grain size is more than 4 or 5 MB. You can easily reproduce that by setting up an Wordpress grain and adding only few content like some images. I was nearly never able to backup such a grain. I also don't see any logs in the admin panel.

This is happening with all kind of browsers. All systems are setup with Debian Jessie.

**Comment by kentonv** [comment]
I've been unable to reproduce this either on Oasis or on a local test server. :/

@JamborJan does this happen for you on Oasis?

**Comment by JamborJan** [comment]
Hey @kentonv I'm not able to reproduce it on Oasis. I'm also no longer able to reproduce it on one of my Sandstorm instances but on a second one with a slower internet connection. The second one is my development and test server sitting in the basement behind a consumer ISP connection. I see that the download of the backup is slow compared to Oasis and my other production system. After few seconds it's decreasing towards 0 and then there is a message "The network connection has been interrupted" in the browsers download view.

Might it be that there is an issue when the download of the backup is not happening at a certain speed or small disconnects occur?

**Comment by kentonv** [comment]
Hmm, my best guess here is that the ISP is doing some traffic shaping -- that would explain the download (or upload from the ISP's point of view) being fast initially and then slowing down, then eventually dropping.

In other words, the connection drops are a network problem, not a Sandstorm problem.

That said, connection drops are pretty common on the internet and any kind of bulk transfer protocol is supposed to have a way to recover with a new connection. In the case of HTTP, that mechanism is range requests. Unfortunately our backup-downloading code currently does not support range requests, hence any connection drop ends the download. My guess is if we added range request support, this problem would be solved.

**Comment by Michael-S** [comment]
I'm not sure if I'm having the same issue or another one, but when I backup grains from the sandstorm server I'm running in my house it times out with the 2GB music collection from Groove Basin.  I've included a section of my sandstorm.log with the error.  (If this is unrelated, let me know and I can open another issue or attach this to a different existing issue.)   The server has 32GB of RAM and is mostly idle aside from Sandstorm when I'm trying to get the backup.  If it matters, the server actually connects over wifi (long story).
[sandstorm_log_fragment.txt](https://github.com/sandstorm-io/sandstorm/files/673520/sandstorm_log_fragment.txt)

**Comment by kentonv** [comment]
@Michael-S Looks like a separate bug. Could you file a new issue? Thanks!

**Comment by ocdtrekkie** [comment]
I just helped another user determine this was likely their issue.

**Comment by xet7** [comment]
In that logfile from @Michael-S is this error:

```
capnp/rpc.c++:142: info: returning failure over rpc; exception = sandstorm/backend.c++:89: failed: expected !isRetry; retry supervisor startup logic failed
```

So it gets disconnected about here:
https://github.com/sandstorm-io/sandstorm/blob/master/src/sandstorm/backend.c%2B%2B#L85-L95

```c++
        if (exception.getType() == kj::Exception::Type::DISCONNECTED) {
          // Oops, disconnected. onDisconnect() should have already fired causing the RunningGrain
          // to unregister itself. Give it an extra turn using evalLater() just in case, then
          // re-run.
          KJ_ASSERT(!isRetry, "retry supervisor startup logic failed");
          return kj::evalLater([=]() mutable {
            return bootGrain(grainId, packageId, command, isNew, devMode, mountProc, true);
          });
        } else {
          return kj::mv(exception);
}
```

@kentonv

Is this:
a) error in that disconnection logic? Is backup streamed as download to user in the fly?
b) issue of not enough disk space?

**Comment by xet7** [comment]
@JamborJan @ocdtrekkie @kentonv 

After upgrading to new experimental WordPress version created by @JamborJan , I am now able to download and upload my WordPress grain that has size 400 MB visible in Sandstorm, and grain .zip size is 388,9 MB. I have Sandstorm installed to Ubuntu 17.10 running on Google Cloud VM that has 1689 MB RAM.

Has somebody fixed this issue? It works for me now. It did not work previously long time ago.

**Comment by ocdtrekkie** [comment]
I don't think it is fixed, but it always worked as long as your download wasn't interrupted in any way. Your connection could be more stable right now.

**Comment by JamborJan** [comment]
I did not fix it in terms of doing stuff on purpose but switching to the latest Wordpress version fixed a lot of compatibility and upstream issues which might also result in the fact that it is working now.
